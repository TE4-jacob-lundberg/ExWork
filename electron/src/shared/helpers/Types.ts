export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  id: string;
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}
