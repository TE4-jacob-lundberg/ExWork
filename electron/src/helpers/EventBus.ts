/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
