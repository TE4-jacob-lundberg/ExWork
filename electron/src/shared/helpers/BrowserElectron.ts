// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
