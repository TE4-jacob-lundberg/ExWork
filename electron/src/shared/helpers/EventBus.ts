/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
