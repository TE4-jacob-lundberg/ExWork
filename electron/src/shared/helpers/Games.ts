import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
