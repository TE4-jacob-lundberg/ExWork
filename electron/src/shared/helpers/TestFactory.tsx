/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
