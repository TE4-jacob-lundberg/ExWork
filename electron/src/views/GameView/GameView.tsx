import React, { useState, useEffect, ReactNode } from 'react';
import { useParams } from 'react-router-dom';
import Styled from '@emotion/styled';
import { useIndexedDB } from 'react-indexed-db';
import { css } from '@emotion/core';
import { Formik, Field, Form } from 'formik';

import LinkCollectionComponent from './components/LinkCollectionComponent';
import PageTitleComponent from '../../shared/components/PageTitleComponent';
import { IGame, ILink } from '../../shared/helpers/Types';
import ButtonComponent from '../../shared/components/ButtonComponent';
import ModalComponent from '../../shared/components/ModalComponent';

const ContainerStyled = Styled.div`
  height: 100%;
  width: 100%;
  position: relative;
  display: flex;
  flex-direction: column;
`;

interface Props {}

interface Params {
  gameID: string;
}

interface IModalContent {
  id: string;
  title: string;
  content: ReactNode;
}

interface IFormValues {
  url?: string;
  label?: string;
  name?: string;
}

const GameView: React.FC<Props> = function () {
  const params = useParams<Params>();
  const [gameData, setGameData] = useState<IGame>();
  const [modalContent, setModalContent] = useState<IModalContent>();
  const db = useIndexedDB('games');

  useEffect(() => {
    if (gameData) return;
    db.getByID(params.gameID).then(resp => setGameData(resp));
  }, [db, params.gameID, gameData]);

  const BackgroundStyled = Styled.div`
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: -1;
    filter: grayscale(50%);
    background: url(${gameData && gameData.image!.url }) 50% ${gameData ? gameData.image!.bannerPos.y : '0'} no-repeat;
    background-size: cover;
  `;

  function handleLinkSubmit(rawValues: IFormValues): void {
    if (!gameData) return;
    const newValues: ILink = Object.assign(rawValues);
    if (!rawValues.url?.match(/$(https:\/\/|http:\/\/)/)) newValues.url = `https://${rawValues.url}`;

    const updatedData = {
      ...gameData,
      links: [
        ...gameData?.links,
        newValues,
      ],
    };
    db.update(updatedData)
      .then(() => {
        setModalContent(undefined);
        setGameData(updatedData);
      })
      .catch(err => console.error(err)); 
  }

  const ActionContainerStyled = Styled.section`
    width: 200px;
    height: 50px;
    background: #222;
    margin-top: auto;
    display: flex;
    justify-content: space-around;
    align-items: center;
  `;

  const ButtonStyled = css`
    width: 100%;
    height: 100%;
    border-radius: 0;
  `;

  const FormStyling = Styled.span`
    & + form {
      width: 50%;
      height: 300px;
      display: flex;
      margin: auto;
      flex-direction: column;

      & > input {
        margin-bottom: 12px;

        &:focus {
          outline: none;
        }
      }
    }
  `;

  const LabelStyled = Styled.label`
    font-size: 1rem;
  `;

  const FormButton = (
    <ButtonComponent
      type="submit"
      onClick={(): string => ''}
      background="green"
      styling={css`
        font-size: 1.5rem;
        margin: auto;
      `}
    >
      Save
    </ButtonComponent>
  );

  return (
    <>
      <ContainerStyled>
        {gameData && (
          <>
            <PageTitleComponent title={gameData.title} />
            <BackgroundStyled />
            <LinkCollectionComponent links={gameData.links} />
            <ActionContainerStyled>
              <ButtonComponent
                onClick={(): void => setModalContent({id: 'link', title: 'Add link', content: (
                  <Formik
                    initialValues={{url: '', label: ''}}
                    onSubmit={(values: IFormValues): void => handleLinkSubmit(values)}
                  >
                    <Form>
                      <LabelStyled htmlFor="url">URL</LabelStyled>
                      <Field name="url" id="url" />

                      <LabelStyled htmlFor="label">Label</LabelStyled>
                      <Field name="label" id="label" />
                      {FormButton}
                    </Form>
                  </Formik>
                )})}
                styling={ButtonStyled}
                data-testid="add-link"
              >
                <i className="material-icons">
                  add
                </i>
              </ButtonComponent>
              <ButtonComponent
                onClick={(): void => setModalContent({id: 'folder', title: 'Create folder', content: (
                  <Formik
                    initialValues={{name: ''}}
                    onSubmit={(values): void => handleLinkSubmit(values)}
                  >
                    <Form>
                      <LabelStyled htmlFor="name">name</LabelStyled>
                      <Field name="name" />
                      {FormButton}
                    </Form>
                  </Formik>
                )})}
                styling={ButtonStyled}
                disabled
              >
                <i className="material-icons">
                  create_new_folder
                </i>
              </ButtonComponent>
            </ActionContainerStyled>
          </>
        )}
      </ContainerStyled>
      {modalContent && (
        <ModalComponent
          title={modalContent.title}
          onClose={(): void => setModalContent(undefined)}
          noActions
        >
          <FormStyling />
          {modalContent.content}
        </ModalComponent>
      )}
    </>
  );
};

export default GameView;
