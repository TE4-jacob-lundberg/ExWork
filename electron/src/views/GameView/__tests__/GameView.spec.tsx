import React from 'react';
import { routerFactory } from '../../../shared/helpers/TestFactory';

import GameView from '../GameView';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: (): object => ({
    gameID: 'csgo',
  }),
}));

const testComponent = <GameView />;

describe('GameView.tsx', () => {
  it('renders correctly', () => {
    const { container } = routerFactory(testComponent);
    
    expect(container).toMatchSnapshot();
  });
});
