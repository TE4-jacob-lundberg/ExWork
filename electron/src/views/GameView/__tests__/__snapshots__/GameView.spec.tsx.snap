// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameView.tsx renders correctly 1`] = `
<div>
  <div
    class="css-1bxr3a5"
  >
    <div
      class="css-j99gk8"
      data-testid="page-title"
    >
      <h1
        class="css-p1pfmz"
      >
        testTitle
      </h1>
    </div>
    <div
      class="css-1xttfd9"
    />
    <div
      class="css-ubcm2u"
    >
      <div
        class="css-cfdpzz"
      >
        <i
          class="material-icons css-dyscyh"
        >
          remove_circle
        </i>
        <button
          class="css-uxkr57"
        >
          testWiki
        </button>
      </div>
    </div>
    <section
      class="css-1lzwb5o"
    >
      <button
        class="css-17epmtx"
        data-testid="add-link"
      >
        <i
          class="material-icons"
        >
          add
        </i>
      </button>
      <button
        class="css-equvm"
        disabled=""
      >
        <i
          class="material-icons"
        >
          create_new_folder
        </i>
      </button>
    </section>
  </div>
</div>
`;
